src/util/Exception.cxx

```diff
@@ -1,99 +1,214 @@
+    return "Privileged instruction";
+  case EXCEPTION_IN_PAGE_ERROR:
+    return "In page error";
+  case EXCEPTION_ILLEGAL_INSTRUCTION:
+    return "Illegal instruction";
+  case EXCEPTION_NONCONTINUABLE_EXCEPTION:
+    return "Noncontinuable exception";
+  case EXCEPTION_STACK_OVERFLOW:
+    return "Stack overflow";
+  case EXCEPTION_INVALID_DISPOSITION:
+    return "Invalid disposition";
+  case EXCEPTION_GUARD_PAGE:
+    return "Guard page";
+  case EXCEPTION_INVALID_HANDLE:
+    return "Invalid handle";
   }
-  case SIGSEGV:
-  {
-    LOG(LOG_ERROR) << "Segmentation Fault";
-    break;
-  }
-  default:
+  return "Unknown exception";
+}
+
+static void LwSystemError(const char *title, const char *text)
+{
+  MessageBox(NULL, text, title, MB_OK | MB_ICONERROR);
+}
+
+static void LwPrintStacktrace(LPEXCEPTION_POINTERS e)
+{
+  char crash_info[256];
+  int filelineinfo_ok = 0;
+  PIMAGEHLP_SYMBOL64 pSym;
+  STACKFRAME sf;
+  DWORD64 dwModBase, Disp64;
+  DWORD Disp;
+  BOOL more = FALSE;
+  IMAGEHLP_LINE64 line;
+  const char *filename = NULL;
+  unsigned int linenum = 0;
+  line.SizeOfStruct = sizeof(IMAGEHLP_LINE64);
+  int count = 0;
+  char modname[MAX_PATH];
+  pSym = (PIMAGEHLP_SYMBOL64)GlobalAlloc(GMEM_FIXED, 16384);
+  ZeroMemory(&sf, sizeof(sf));
+  sf.AddrPC.Mode = AddrModeFlat;
+  sf.AddrStack.Mode = AddrModeFlat;
+  sf.AddrFrame.Mode = AddrModeFlat;
+  sf.AddrReturn.Mode  = AddrModeFlat;
+
+  PCONTEXT context = e->ContextRecord;
+  DWORD machine_type = 0;
+  sf.AddrPC.Offset = context->Rip;
+  sf.AddrFrame.Offset = context->Rbp;
+  sf.AddrStack.Offset = context->Rsp;
+  machine_type = IMAGE_FILE_MACHINE_AMD64;
+
+  // Record exception info
+  LOG(LOG_ERROR) << "Exception [" << LwPrintExceptionName(e->ExceptionRecord->ExceptionCode) << "] 0x" << std::hex << (unsigned int) e->ExceptionRecord->ExceptionCode;
+  LOG(LOG_ERROR) << "Exception Address: 0x" << std::hex << e->ExceptionRecord->ExceptionAddress;
+
+  // Record stacktrace
+  LOG(LOG_ERROR) << "Stacktrace:";
+
+  while (1)
   {
-    LOG(LOG_ERROR) << "Unknown signal received";
-    break;
-  }
+    more = StackWalk(machine_type, GetCurrentProcess(), GetCurrentThread(), &sf, context, NULL, SymFunctionTableAccess64, SymGetModuleBase64, NULL);
+    if (!more || sf.AddrFrame.Offset == 0)
+    {
+      break;
+    }
+    dwModBase = SymGetModuleBase64(GetCurrentProcess(), sf.AddrPC.Offset);
+    if (dwModBase)
+    {
+      GetModuleFileName((HINSTANCE) dwModBase, modname, MAX_PATH);
+    }
+    else
+    {
+      strcpy(modname, "Unknown");
+    }
+
+    pSym->SizeOfStruct = sizeof(IMAGEHLP_SYMBOL);
+    pSym->MaxNameLength = MAX_PATH;
+
+    // unwind callstack
+    if (SymGetSymFromAddr64(GetCurrentProcess(), sf.AddrPC.Offset, &Disp64, pSym))
+    {
+      filelineinfo_ok = SymGetLineFromAddr64(GetCurrentProcess(), sf.AddrPC.Offset, &Disp, &line);
+      filename = !!filelineinfo_ok ? line.FileName : "UNKNOWN FILE";
+      linenum = !!filelineinfo_ok ? line.LineNumber : 0;
```

Comment:
why double negation `!!filelineinfo_ok`

Reply:
fileinfo_ok is int, we dont known what return function here, that correctly convert positive int in to bool usually used !!
what happens int a = 1; !a -> 0; !0 -> true; !!a (1) -> true

Reply:
Clear, although I would rather test with (filelineinfo_ok != 0)